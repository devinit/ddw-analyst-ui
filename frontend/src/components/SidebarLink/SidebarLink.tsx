import * as React from 'react';
import { RouteComponentProps, withRouter } from 'react-router-dom';

export interface SidebarLinkProps extends RouteComponentProps<{}> {
  icon?: string;
  to: string;
  root?: boolean;
  textMini?: string;
  textNormal?: string;
  caret?: boolean;
  single?: boolean;
  onClick?: (to: string) => void;
  [`data-testid`]?: string;
}

class SidebarLink extends React.Component<SidebarLinkProps> {
  static defaultProps: Partial<SidebarLinkProps> = {
    caret: false
  };

  render() {
    return (
      <a
        className="nav-link"
        data-toggle={ this.props.root && !this.props.single ? 'collapse' : '' }
        href={ this.props.to }
        onClick={ this.onClick }
        data-testid={ this.props['data-testid'] }
      >
        { this.props.icon ? <i className="material-icons">{ this.props.icon }</i> : null }
        { this.renderContent() }
      </a>
    );
  }

  private renderContent() {
    if (this.props.root || this.props.single) {
      return (
        <p> { this.props.textNormal || this.props.textMini }
          { this.props.caret ? <b className="caret"/> : null }
        </p>
      );
    }

    return (
      <React.Fragment>
        <span className="sidebar-mini"> { this.props.textMini } </span>
        <span className="sidebar-normal"> { this.props.textNormal } </span>
      </React.Fragment>
    );
  }

  private onClick = (event: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => {
    event.preventDefault();
    if (this.props.onClick) {
      this.props.onClick(this.props.to);
    }
    if (this.props.to.indexOf('#') === -1) {
      this.props.history.push(this.props.to);
    }
  }
}

const RouteContainer = withRouter(SidebarLink);

export { RouteContainer as SidebarLink };
