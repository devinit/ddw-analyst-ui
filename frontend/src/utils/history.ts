/* eslint-disable @typescript-eslint/naming-convention */
import axios, { AxiosResponse } from 'axios';
import * as localForage from 'localforage';
import { api, localForageKeys } from '.';
import { FrozenData } from '../components/SourceHistoryListItem/utils';
import { SavedQueryData } from '../components/DatasetHistoryCard/utils/types';

const SOURCE_HISTORY_BASEPATH = api.routes.FETCH_SOURCE_HISTORY;
const OPERATION_HISTORY_BASEPATH = api.routes.FETCH_DATASET_HISTORY;

export interface QueryResult<T = FrozenData> {
  count: number;
  next: number | null;
  previous: number | null;
  results: Array<T>;
}

interface FetchOptions {
  limit?: number;
  offset?: number;
}

export const fetchDataSourceHistory = async (
  id: number,
  options: FetchOptions = { limit: 10, offset: 0 },
): Promise<AxiosResponse<QueryResult>> => {
  const token = await localForage.getItem<string>(localForageKeys.API_KEY);
  const headers = {
    'Content-Type': 'application/json',
    Authorization: `token ${token}`,
  };

  return await axios(
    `${SOURCE_HISTORY_BASEPATH}${id}?limit=${options.limit || 10}&offset=${options.offset || 0}`,
    { headers },
  );
};

export const fetchFrozenData = async (id: number): Promise<AxiosResponse<FrozenData>> => {
  const token = await localForage.getItem<string>(localForageKeys.API_KEY);
  const headers = {
    'Content-Type': 'application/json',
    Authorization: `token ${token}`,
  };

  return await axios(`${api.routes.FROZEN_DATA}${id}`, { headers });
};

export const createFrozenData = async (data: FrozenData): Promise<AxiosResponse> => {
  const token = await localForage.getItem<string>(localForageKeys.API_KEY);
  const response = await axios.request({
    url: api.routes.FROZEN_DATA,
    method: 'post',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `token ${token}`,
    },
    data,
  });

  return response;
};

export const deleteFrozeData = async (id: number): Promise<AxiosResponse> => {
  const token = await localForage.getItem<string>(localForageKeys.API_KEY);
  const response = await axios.request({
    url: `${api.routes.FROZEN_DATA}${id}`,
    method: 'delete',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `token ${token}`,
    },
  });

  return response;
};

export const fetchOperationHistory = async (
  id: number,
  options: FetchOptions = { limit: 10, offset: 0 },
): Promise<AxiosResponse<QueryResult<SavedQueryData>>> => {
  const token = await localForage.getItem<string>(localForageKeys.API_KEY);
  const headers = {
    'Content-Type': 'application/json',
    Authorization: `token ${token}`,
  };

  return await axios(
    `${OPERATION_HISTORY_BASEPATH}${id}?limit=${options.limit || 10}&offset=${options.offset || 0}`,
    { headers },
  );
};

export const fetchSavedQueryData = async (id: number): Promise<AxiosResponse<SavedQueryData>> => {
  const token = await localForage.getItem<string>(localForageKeys.API_KEY);
  const headers = {
    'Content-Type': 'application/json',
    Authorization: `token ${token}`,
  };

  return await axios(`${api.routes.SINGLE_SAVED_QUERYSET}${id}`, { headers });
};

export const createSavedQueryData = async (data: SavedQueryData): Promise<AxiosResponse> => {
  const token = await localForage.getItem<string>(localForageKeys.API_KEY);
  const response = await axios.request({
    url: api.routes.SAVED_QUERYSETS,
    method: 'post',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `token ${token}`,
    },
    data,
  });

  return response;
};

export const deleteSavedQueryData = async (id: number): Promise<AxiosResponse> => {
  const token = await localForage.getItem<string>(localForageKeys.API_KEY);
  const response = await axios.request({
    url: `${api.routes.SINGLE_SAVED_QUERYSET}${id}`,
    method: 'delete',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `token ${token}`,
    },
  });

  return response;
};
