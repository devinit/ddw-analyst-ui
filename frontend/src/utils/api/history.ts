import axios, { AxiosResponse } from 'axios';
import * as localForage from 'localforage';
import { localForageKeys, api } from '..';
import { FrozenData } from '../../components/SourceHistoryListItem/utils';

const BASEPATH = api.routes.FETCH_SOURCE_HISTORY;

export interface ResultData {
  id: number;
  name: string;
  updated_on: string;
  is_draft: boolean;
  operation_query: string;
}

export interface QueryResult {
  data: Array<ResultData>;
}

interface FetchOptions {
  limit?: number;
  offset?: number;
}

export const fetchDataSourceHistory = async (
  id: number,
  options: FetchOptions = { limit: 10, offset: 0 },
): Promise<QueryResult> => {
  const token = await localForage.getItem<string>(localForageKeys.API_KEY);
  const headers = {
    'Content-Type': 'application/json',
    Authorization: `token ${token}`,
  };

  return await axios(
    `${BASEPATH}${id}?limit=${options.limit || 10}&offset=${options.offset || 0}`,
    { headers },
  );
};

export const createFrozenData = async (data: FrozenData): Promise<AxiosResponse> => {
  const token = await localForage.getItem<string>(localForageKeys.API_KEY);
  const response = await axios.request({
    url: api.routes.FROZEN_DATA,
    method: 'post',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `token ${token}`,
    },
    data,
  });

  return response;
};
