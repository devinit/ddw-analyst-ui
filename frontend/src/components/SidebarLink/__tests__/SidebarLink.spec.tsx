/**
 * @jest-environment jsdom
 */
import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { SidebarLink } from '../SidebarLink';
import { Router } from 'react-router';
import { createBrowserHistory } from 'history';

test('renders the correct link with the default props', () => {
  const renderer = TestRenderer
    .create(
      <Router history={ createBrowserHistory() }>
        <SidebarLink to="link"/>
      </Router>
    )
    .toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders root links correctly', () => {
  const renderer = TestRenderer
    .create(
      <Router history={ createBrowserHistory() }>
        <SidebarLink to="link" root/>
      </Router>
    )
    .toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders single links correctly', () => {
  const renderer = TestRenderer
    .create(
      <Router history={ createBrowserHistory() }>
        <SidebarLink to="link" single/>
      </Router>
    )
    .toJSON();

  expect(renderer).toMatchSnapshot();
});

test('does not render with the data-toggle when configured as single', () => {
  const renderer = TestRenderer
    .create(
      <Router history={ createBrowserHistory() }>
        <SidebarLink to="link" root single/>
      </Router>
    )
    .toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders with a caret when specified', () => {
  const renderer = TestRenderer
    .create(
      <Router history={ createBrowserHistory() }>
        <SidebarLink to="link" caret/>
      </Router>
    )
    .toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders root links with a label when specified', () => {
  const renderer = TestRenderer
    .create(
      <Router history={ createBrowserHistory() }>
        <SidebarLink root to="link" textNormal="Label"/>
      </Router>
    )
    .toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders non-root links with both labels', () => {
  const renderer = TestRenderer
    .create(
      <Router history={ createBrowserHistory() }>
        <SidebarLink to="link" textNormal="Label" textMini="L"/>
      </Router>
    )
    .toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders with an icon when specified', () => {
  const renderer = TestRenderer
    .create(
      <Router history={ createBrowserHistory() }>
        <SidebarLink to="link" icon="home"/>
      </Router>
    )
    .toJSON();

  expect(renderer).toMatchSnapshot();
});
