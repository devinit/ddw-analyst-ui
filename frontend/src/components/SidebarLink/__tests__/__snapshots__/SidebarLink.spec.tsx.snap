// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render with the data-toggle when configured as single 1`] = `
<a
  className="nav-link"
  data-toggle=""
  href="link"
  onClick={[Function]}
>
  <p>
     
  </p>
</a>
`;

exports[`renders non-root links with both labels 1`] = `
<a
  className="nav-link"
  data-toggle=""
  href="link"
  onClick={[Function]}
>
  <span
    className="sidebar-mini"
  >
     
    L
     
  </span>
  <span
    className="sidebar-normal"
  >
     
    Label
     
  </span>
</a>
`;

exports[`renders root links correctly 1`] = `
<a
  className="nav-link"
  data-toggle="collapse"
  href="link"
  onClick={[Function]}
>
  <p>
     
  </p>
</a>
`;

exports[`renders root links with a label when specified 1`] = `
<a
  className="nav-link"
  data-toggle="collapse"
  href="link"
  onClick={[Function]}
>
  <p>
     
    Label
  </p>
</a>
`;

exports[`renders single links correctly 1`] = `
<a
  className="nav-link"
  data-toggle=""
  href="link"
  onClick={[Function]}
>
  <p>
     
  </p>
</a>
`;

exports[`renders the correct link with the default props 1`] = `
<a
  className="nav-link"
  data-toggle=""
  href="link"
  onClick={[Function]}
>
  <span
    className="sidebar-mini"
  >
     
     
  </span>
  <span
    className="sidebar-normal"
  >
     
     
  </span>
</a>
`;

exports[`renders with a caret when specified 1`] = `
<a
  className="nav-link"
  data-toggle=""
  href="link"
  onClick={[Function]}
>
  <span
    className="sidebar-mini"
  >
     
     
  </span>
  <span
    className="sidebar-normal"
  >
     
     
  </span>
</a>
`;

exports[`renders with an icon when specified 1`] = `
<a
  className="nav-link"
  data-toggle=""
  href="link"
  onClick={[Function]}
>
  <i
    className="material-icons"
  >
    home
  </i>
  <span
    className="sidebar-mini"
  >
     
     
  </span>
  <span
    className="sidebar-normal"
  >
     
     
  </span>
</a>
`;
